design/control-data-plane-separation/design.md

```diff
@@ -23,377 +23,118 @@ architecture does not allow the control plane and data plane to scale independen
 
-    // request connection to a management plane, A → C
-    AgentConnectRequest agent_connect_request = 6 [(gogoproto.jsontag) = "agent_connect_request"];
+`FileService`: when the agent receives a `ConfigApplyRequest`, it sends a `GetFile` request to this service to download the file contents contained in the original `FileOverview`.
 
-    // connection response to the data plane, C → A
-    AgentConnectResponse agent_connect_response = 7 [(gogoproto.jsontag) = "agent_connect_response"];
+Some API methods will not need to be implemented. For example, `UpdateFile` is used by the agent to send a file that was updated on the agent side to the control plane. In our case, this won't happen since our control plane has full control over the config. Methods relating to this functionality can be stubbed out.
 
-    // request Configuration parameters for agent, A → C
-    AgentConfigRequest agent_config_request = 8 [(gogoproto.jsontag) = "agent_config_request"];
+### Metrics
 
-    // configuration parameters for Agent C → A. This message can be sent asynchronously as well
-    AgentConfig agent_config = 9 [(gogoproto.jsontag) = "agent_config"];
+Agent can be configured to expose metrics on a `/metrics` endpoint. Our control plane will need to configure agent to do this so that Prometheus can scrape each agent for nginx metrics.
 
-    // DataplaneStatus reports Dataplane metrics the Agent is aware of
-    DataplaneStatus dataplane_status = 11 [(gogoproto.jsontag) = "dataplane_status"];
+### NGINX Plus
 
-    // EventReport reports events the Agent is aware of, e.g. Start/Stop of Agent, Config Apply NGINX
-    f5.nginx.agent.sdk.events.EventReport event_report = 12 [(gogoproto.jsontag) = "event_report"];
+#### Upstream server state
 
-    // DataplaneSoftwareDetails contains details for additional software running on the dataplane that pertains
-    // to NGINX Agent
-    DataplaneSoftwareDetails dataplane_software_details = 13 [(gogoproto.jsontag) = "dataplane_software_details"];
+In the current implementation using NGINX Plus, when only upstream servers change, NGF writes the upstream servers in the nginx config, but does not reload. It then calls the NGINX Plus API to update the servers. This process allows us to update the upstream servers using the API and not have to reload nginx, while still having the upstream servers exist in the nginx config for easy debugging and consistency. However, when using agent, any config write will result in a reload. To preserve the ability to update upstreams with the API without needing a reload, we'll have to utilize a `state` file instead of writing the servers directly in the nginx config. This way the list of servers is still available on the filesystem for debugging, but is written by the nginx when making the API call instead of by the control plane directly.
```

Comment:
I thought there was a plan to expose N+ API calls through the agent API. Has that changed? How does the agent know that the file we are sending is a `state` file and it shouldn't trigger a reload?

Reply:
We don't write that state file, nginx writes it when we call the API.